File: revisit/first_missing_positive.js
// https://leetcode.com/problems/first-missing-positive/

/**
 * @param {number[]} nums
 * @return {number}
 */
var firstMissingPositive = function (nums) {
	// O(n) time, constant space
	for (let i = 0; i < nums.length; i++) {
		if (nums[i] < 0) nums[i] = 0;
	}

	for (let i = 0; i < nums.length; i++) {
		const index = Math.abs(nums[i]) - 1;
		if (index < 0 || index >= nums.length) continue;
		if (nums[index] > 0) nums[index] = -nums[index];
		if (nums[index] == 0) nums[index] = -(nums.length + 1);
	}

	for (let i = 0; i < nums.length; i++) {
		if (nums[i] >= 0) {
			return i + 1;
		}
	}

	return nums.length + 1;

	// O(n) time and space solution
	// const found = [undefined];
	// found[nums.length] = undefined

	// for (const num of nums) {
	//     if (num > 0 && num <= nums.length) {
	//         found[num - 1] = true;
	//     }
	// }

	// for (let i = 0; i < found.length; i++) {
	//     if (found[i] === undefined) return i + 1
	// }

	// return nums.length

	// Sort, then linear scan: O(nlogn)
	// nums.push(0)
	// nums.sort((a, b) => a - b)

	// for (let i = 0; i < nums.length ; i++) {
	//    const left = nums[i]
	//    const right = i < nums.length-1 ? nums[i+1] : Number.MAX_SAFE_INTEGER

	//    if (left >= 0 && left+1 < right) {
	//     return left+1
	//    }
	// }

	// Naive solution: O(n^2)
	// for (let i = 1; i <= nums.length+1; i++) {
	//     let found = false;
	//     for (let j = 0; j < nums.length; j++) {
	//         if (nums[j] == i) {
	//             found = true
	//             break
	//         }
	//     }
	//     if (!found) {
	//         return i
	//     }
	// }
};

File: src/python/finished/first_missing_positive.py
from typing import List

# https://leetcode.com/problems/first-missing-positive/

class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        N = len(nums)
        # Set all negative values to 0 since they are irrelevant
        for i in range(N):
            nums[i] = max(nums[i], 0)
        
        # For each positive element k, we signal that k is present by
        # setting (k-1) th element as negative without changing its absolute value
        for i in range(N):
            k_minus_1 = abs(nums[i]) - 1
            # Verify k is positive and between 1 to N
            if k_minus_1 < 0 or k_minus_1 >= N:
                # In an array of length N, we can at most cover up to N from 1.
                # Any value above this range is irrelevant
                continue
            if nums[k_minus_1] > 0:
                # We mark k as present
                nums[k_minus_1] = -nums[k_minus_1]
            if nums[k_minus_1] == 0:
                # 0 is a neutral number
                # We set (k-1)th element to a large negative value
                # If we set nums[k_minus_1] to a value m smaller than (N+1), we might 
                # mislead the algorithm into thinking that m exists when it may not.
                nums[k_minus_1] = -(N+1) # Any value larger than N+1 will work

        # Find the first index, l = k-1, that has a positive element.
        # A positive element indictes that k is not yet seen. This is our result
        for i in range(N):
            if nums[i] >= 0:
                # (k-1) = i, k = i + 1
                return i + 1
        return N+1